server/src/main/java/io/druid/server/BrokerQueryResource.java

```diff
@@ -0,0 +1,167 @@
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied. See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package io.druid.server;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.fasterxml.jackson.jaxrs.smile.SmileMediaTypes;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import com.google.inject.Inject;
+import com.metamx.common.guava.Sequences;
+import com.metamx.emitter.service.ServiceEmitter;
+import io.druid.client.TimelineServerView;
+import io.druid.client.selector.ServerSelector;
+import io.druid.common.utils.JodaUtils;
+import io.druid.guice.annotations.Json;
+import io.druid.guice.annotations.Smile;
+import io.druid.query.DataSource;
+import io.druid.query.LocatedSegmentDescriptor;
+import io.druid.query.Query;
+import io.druid.query.QuerySegmentWalker;
+import io.druid.query.SegmentDescriptor;
+import io.druid.query.TableDataSource;
+import io.druid.server.coordination.DruidServerMetadata;
+import io.druid.server.initialization.ServerConfig;
+import io.druid.server.log.RequestLogger;
+import io.druid.timeline.TimelineLookup;
+import io.druid.timeline.TimelineObjectHolder;
+import io.druid.timeline.partition.PartitionChunk;
+import org.joda.time.Interval;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ */
+@Path("/druid/v2/")
+public class BrokerQueryResource extends QueryResource
+{
+  private final TimelineServerView brokerServerView;
+
+  @Inject
+  public BrokerQueryResource(
+      ServerConfig config,
+      @Json ObjectMapper jsonMapper,
+      @Smile ObjectMapper smileMapper,
+      QuerySegmentWalker texasRanger,
+      ServiceEmitter emitter,
+      RequestLogger requestLogger,
+      QueryManager queryManager,
+      TimelineServerView brokerServerView
+  )
+  {
+    super(config, jsonMapper, smileMapper, texasRanger, emitter, requestLogger, queryManager);
+    this.brokerServerView = brokerServerView;
+  }
+
+  @POST
+  @Path("/candidates")
+  @Produces({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE})
+  @Consumes({MediaType.APPLICATION_JSON, SmileMediaTypes.APPLICATION_JACKSON_SMILE, APPLICATION_SMILE})
+  public Response getQueryTargets(
+      InputStream in,
+      @QueryParam("pretty") String pretty,
+      @Context final HttpServletRequest req // used only to get request content-type and remote address
+  ) throws IOException
+  {
+    final String reqContentType = req.getContentType();
```

Comment:
tried couple of times but it didn't make code any cleaner because it's complex code accessing many parameters. I'll try once more with shorter grip.